src/newTest/java/org/opensearch/test/framework/TestIndex.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2021-2022 floragunn GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.test.framework;
+
+import org.opensearch.action.admin.indices.create.CreateIndexRequest;
+import org.opensearch.action.admin.indices.exists.indices.IndicesExistsRequest;
+import org.opensearch.client.Client;
+import org.opensearch.common.settings.Settings;
+
+public class TestIndex {
+
+    private final String name;
+    private final Settings settings;
+    
+    public TestIndex(String name, Settings settings) {
+        this.name = name;
+        this.settings = settings;
+        
+    }
+
+    public void create(Client client) {
+        if (!client.admin().indices().exists(new IndicesExistsRequest(name)).actionGet().isExists()) {
+            client.admin().indices().create(new CreateIndexRequest(name).settings(settings)).actionGet();
+        }
+    }
+
+    public String getName() {
+        return name;
+    }
+
+
+    public static Builder name(String name) {
+        return new Builder().name(name);
+    }
+
+    public static class Builder {
```

Comment:
Is there a way we can use annotations instead of having to author all of the builders by hand?

Reply:
We reviewed the Immutables library for this. It looks cool. However, the JSON we need to write is not a direct mapping from the builder structures. Thus we came to the conclusion, to do this manually for now.

Reply:
> the JSON we need to write is not a direct mapping from the builder structures

Do you have an example of how this breaks down?

> to do this manually for now.

Is there another way to do this or generate this? Its painful to review or possible deal with bugs in boiler plate functionality.

Reply:
An example for the data structures:

In the security plugin configuration, there is an N:M relation between users and roles:

`internal_users.yml`:

```
demo_user:
  opendistro_security_roles:
  - role_1
```
`roles.yml`:

```
demo_user_role:
  cluster_permissions:
    ...
```


However, for the testing purposes, we would like to keep the configuration as simple as possible:

```
TestSecurityConfig.User DEMO_USER = new TestSecurityConfig.User("demo_user").roles(new Role("role_1").clusterPermissions("*"));
```
Of course, sometimes more complex role definitions are necessary. But in the vast majority, it is sufficient to have exactly one role per user and only one user per role.

In order to achieve this, we have to collect all roles of the defined users and aggregate this - with a scoped name - in the `roles.yml` file. 

Thus, we need quite elaborate processing of these structures. Thus, at least the direct JSON mapping feature of Immutables cannot be used. Of course, we can add further code which takes the immutable object trees and builds separate JSON trees from it. However, I am not sure if this will be more simple.

Reply:
Thanks for the explanation, it seems like we are building objects with the structure of the product in mind, I am considering this from the structure of a test cases, which would have is declarative builders for the test configuration that would use some kind of `TestConfigApplier` map them onto the product's structure as needed.

By separating these concerns it might simplify test authorship, what do you think?

Reply:
Should that `TestConfigApplier` be transparent or visible for the actual test code?

Reply:
There are two classes of tests, those that just need any kind of configuration whereafter they build on that configuration to set up the correct test, in the other case, there are deep tests of how that configuration is authored and used.

Those deep tests would seem to be more specific so the actual tooling isn't as important to those usages, where the first case is the more common one.  At this point, I think it would be useful to merge and we can examine improving on this structure in future pull requests.